src/sonic_ax_impl/mibs/ietf/rfc2737.py

```diff
@@ -35,76 +110,256 @@ def __init__(self):
         self.statedb = SonicV2Connector()
         self.statedb.connect(self.statedb.STATE_DB)
 
-        # List of available sub IDs.
-        self.physical_classes = []
+        self.if_alias_map = {}
+
+        # List of available sub OIDs.
+        self.physical_entities = []
+
         # Map sub ID to its data.
         self.physical_classes_map = {}
+        self.physical_description_map = {}
+        self.physical_hw_version_map = {}
+        self.physical_serial_number_map = {}
+        self.physical_mfg_name_map = {}
+        self.physical_model_name_map = {}
 
     def reinit_data(self):
         """
         Re-initialize all data.
         """
-        device_metadata = self.statedb.get_all(self.statedb.STATE_DB, self.DEVICE_METADATA)
 
-        # TODO: Add support for unicode in swsssdk
-        chassis_serial_number = ""
+        # reinit cache
+        self.physical_classes_map = {}
+        self.physical_description_map = {}
+        self.physical_hw_version_map = {}
+        self.physical_serial_number_map = {}
+        self.physical_mfg_name_map = {}
+        self.physical_model_name_map = {}
+
+        device_metadata = mibs.get_device_metadata(self.statedb)
+        chassis_sub_id = (self.CHASSIS_ID, )
+        self.physical_entities = [chassis_sub_id]
+
         if not device_metadata or not device_metadata.get(b"chassis_serial_number"):
             chassis_serial_number = ""
         else:
             chassis_serial_number = device_metadata[b"chassis_serial_number"]
-            self.physical_classes = [(self.CHASSIS_ID, )]
-            self.physical_classes_map = {
-                    (self.CHASSIS_ID, ): (PhysicalClass.CHASSIS, chassis_serial_number.decode("utf-8"))
-                }
 
+        self.physical_classes_map[chassis_sub_id] = PhysicalClass.CHASSIS
+        self.physical_serial_number_map[chassis_sub_id] = chassis_serial_number
+
+        # update interface maps
+        _, self.if_alias_map, _, _, _ = \
+            mibs.init_sync_d_interface_tables(SonicV2Connector())
+            
     def update_data(self):
```

Comment:
>update_data [](start = 8, length = 11)

This function is time critical. Try:
1. move all constants to __init__: like transceiver_key_pattern
2. move all semi constants to reinit_data: like transceiver_entries

Reply:
1. Done, transceiver_key_pattern now is a constant class member
2. I agree that data in transceiver_entries like serial_num, vendor name should not change until we plug out the cable(which is quite rare), so having it in update_data() routine is not a best idea, however if I place it in reinit_data() we will receive changes about cable plugged/unplugged only every 1 min. On the other side, sensor values are dynamic data, so they are updated in update_data() every 5 sec. So we will have an interval when tables are not in sync. (E.g Ethernet0 is plugged out - 1 min there is still Ethernet0 in entities MIB, however Ethernet0 temperature sensor is not present in sensor MIB anymore)
Another way I thought about is that we can fetch the list of transceivers from DB in update_data(). If fetched list is the same as in last update_data() call - no update is needed, otherwise update cached values. In this case there is also an issue - If plug in and plug out happens in this 5 sec interval between update_data() calls we will not capture any changes and will have wrong values in cache until next reinit_data() call.

Reply:
Thanks for the thoughtful analysis. I prefer your 2nd, please keep explorer a better way to poll events, it will lead to considerable CPU optimization.

This optimization would benefit many other OIDs as well.

---
In reply to: [200943979](https://github.com/Azure/sonic-snmpagent/pull/69#discussion_r200943979) [](ancestors = 200943979)

Reply:
I refactored the code in update_data(). reinit_data() fully reinits the cache. In update_data() it listens to changes in STATE DB TRANSCEIVER_INFO table and updates only in case of change. In case of SET event we update only one entry, in case of DELETE event we remove this transceiver from available OIDs (not clearing the cache, cache will be refreshed in next reinit_data() call anyway).
pubsub.get_message() is non-blocking and also returns all unconsumed changes, so I'm not skipping events.